docs/proposals/automatically-propagate-relevant-resources/README.md

```diff
@@ -0,0 +1,181 @@
+---
+title: Automatically Propagate Relevant Resources
+authors:
+- "@mrlihanbo"
+reviewers:
+- "@TBD"
+approvers:
+- "@TBD"
+
+creation-date: 2021-09-06
+
+---
+
+# Automatically Propagate Relevant Resources
+
+## Summary
+
+Secrets or ConfigMaps can be mounted as data volumes or exposed as environment variables to be used by a container in a Pod.
+Meanwhile, users can use a Service to expose the application running on a set of Pods.
+ 
+Thus, it is necessary to deploy the relevant resource into the same member clusters when propagating a Pod by Karmada, i.e, 
+Pod dependencies(ConfigMaps and Secrets) and dependents(Services and Ingresses) should "follow" the delegate pods.
+
+Usually, users don't create Pods directly. Instead, create them using workload resources such as Deployment, StatefulSet or Job.
+
+So I take Deployment as example. The relevant resources include Pod dependencies(ConfigMaps and Secrets) and dependents(Services and Ingresses). 
+
+For now, Karmada provides two ways to propagate relevant resources for a Deployment:
+- The relevant resources share the same propagation policy with the Deployment.
+- Create individual propagation policy for each relevant resource, but users need to make sure the resource scheduled to needed clusters. 
+
+These methods work for most cases, but still has disadvantages:
+- For the first method, the dependencies(config maps and secrets) are bound with one Deployment which means it can't be mounted by other Deployments.
+- For the second method, users should keep watching the schedule result of the Deployment.
+
+This proposal aims to provide a strategy to intelligently propagate relevant resources for workload resources such as Deployment.
+
+## Motivation
+
+Users need to deploy relevant resource manually when deploying Pods or workload resources by Karmada. It is a heavy burden to maintain target clusters for
+relevant resources.
+
+On the other hand, Karmada needs to provide a way to automatically propagate relevant resources to needed clusters.
+
+### Goals
+
+- Provide a strategy to automatically propagate relevant resources of workload resources to needed clusters.
+
+### Non-Goals
+
+- Deprecate the default propagate process for relevant resources.
+ * Relevant resource like ConfigMap, Secret and Service can be propagated by propagation policy as usual.
```

Comment:
This proposal is really useful👍. I can't wait experiencing this feature any more.

By the way, I'm curious about the implementation of finding relevant resource. IMO, for ConfigMap and Secret we find them by parsing field `spec.volume` of workload, and for Service we find them by filtering services whose filed `spec.selector` matches the label. Am I right?

Reply:
> This proposal is really useful👍. I can't wait experiencing this feature any more.
> 
> By the way, I'm curious about the implementation of finding relevant resource. IMO, for ConfigMap and Secret we find them by parsing field `spec.volume` of workload, and for Service we find them by filtering services whose filed `spec.selector` matches the label. Am I right?

Yes, the implementation is like what you said. I will submit the pull request soon so that we can discuss more about it. Besides, I have no idea about automatically propagate ingress which are more complex.

Reply:
Well, ingress is much more complicated. And we should process carefully. The field `ingress.spec.rules.http.paths.backend.serviceName` and `ingress.spec.defaultBackend.service.name` are relevant.